segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1407 @@
+                }
+
+                if (targetSizeAfterConcat + next.getLength() > segmentMetadata.getMaxRollinglength() || next.getLength() > config.getMaxSizeLimitForNativeConcat()) {
+                    break;
+                }
+
+                chunksToConcat.add(new ChunkInfo(next.getLength(), nextChunkName));
+                targetSizeAfterConcat += next.getLength();
+
+                nextChunkName = next.getNextChunk();
+            }
+            // Note - After above while loop is exited nextChunkName points to chunk next to last one to be concat.
+            // Which means target should now point to it as next after concat is complete.
+
+            // If there are chunks that can be appended together then concat them.
+            if (chunksToConcat.size() > 1) {
+                // Concat
+
+                ConcatArgument[] concatArgs = new ConcatArgument[chunksToConcat.size()];
+                for (int i = 0; i < chunksToConcat.size(); i++) {
+                    concatArgs[i] = ConcatArgument.fromChunkInfo(chunksToConcat.get(i));
+                }
+
+                if (!useAppend && chunkStorage.supportsConcat()) {
+                    int length = chunkStorage.concat(concatArgs);
+                } else {
+                    concatUsingAppend(concatArgs);
+                }
+
+                // Set the pointers
+                target.setLength(targetSizeAfterConcat);
+                target.setNextChunk(nextChunkName);
+
+                // If target is the last chunk after this then update metadata accordingly
+                if (null == nextChunkName) {
+                    segmentMetadata.setLastChunk(target.getName());
+                    segmentMetadata.setLastChunkStartOffset(segmentMetadata.getLength() - target.getLength());
+                }
+
+                // Update metadata for affected chunks.
+                for (int i = 1; i < concatArgs.length; i++) {
+                    txn.delete(concatArgs[i].getName());
+                }
+                txn.update(target);
+                txn.update(segmentMetadata);
+            }
+
+            // Move on to next place in list where we can concat if we are done with append based concats.
+            if (!useAppend) {
+                targetChunkName = nextChunkName;
+            }
+
+            // Toggle
+            useAppend = !useAppend;
+        }
+    }
+
+    private void concatUsingAppend(ConcatArgument[] concatArgs) throws ChunkStorageException {
+        long writeAtOffset = concatArgs[0].getLength();
+        val writeHandle = ChunkHandle.writeHandle(concatArgs[0].getName());
+        for (int i = 1; i < concatArgs.length; i++) {
+            int readAtOffset = 0;
+            val arg = concatArgs[i];
+            int bytesToRead = Math.toIntExact(arg.getLength());
+
+            while (bytesToRead > 0) {
+                byte[] buffer = new byte[Math.min(config.getMaxBufferSizeForChunkDataTransfer(), bytesToRead)];
```

Comment:
how big is expected to be this buffer ?
shall we consider reusing the same buffer, with a fixed size ?

Reply:
Currently it is set to be 1MB. (ChunkStorageManagerConfig.maxBufferSizeForChunkDataTransfer).

Reply:
There are several threads and this method should be called rarely. (only called when data is too small for calling native concat eg. for HDFS or ExtendedS3)
Definitely need to  measure and fix it in future if required.